internal/controller/stages/stages.go

```diff
@@ -706,30 +713,42 @@ func (r *reconciler) syncNormalStage(
 		}
 
 		// Initiate or follow-up on verification if required
-		// NOTE: If stage cache is stale, phase can be StagePhaseNotApplicable
-		//       even though current freight is not empty in that case
-		//       check if verification step is necessary and if yes execute
-		//       step irrespective of phase
-		if (status.Phase == kargoapi.StagePhaseVerifying || status.Phase == kargoapi.StagePhaseNotApplicable) &&
-			stage.Spec.Verification != nil {
-			if status.CurrentFreight.VerificationInfo == nil {
-				if status.Health == nil || status.Health.Status == kargoapi.HealthStateHealthy {
-					log.Debug("starting verification")
-					status.CurrentFreight.VerificationInfo = r.startVerificationFn(ctx, stage)
+		if stage.Spec.Verification != nil {
+			// Confirm if a rerun of verification is requested. If so, clear the
+			// verification info and start the verification process again.
+			info := status.CurrentFreight.VerificationInfo
+			if info != nil && info.AnalysisRun != nil && info.Phase.IsTerminal() {
+				if v, ok := stage.GetAnnotations()[kargoapi.AnnotationKeyReconfirm]; ok && v == info.AnalysisRun.Name {
+					logger.Debug("reconfirming verification")
+					status.Phase = kargoapi.StagePhaseVerifying
+					status.CurrentFreight.VerificationInfo = nil
```

Comment:
@krancour this has ended up slightly different than we discussed offline. As requeueing would have caused the annotation to be removed prematurely, making it impossible to determine later on if we should (not) take existing AnalysisRuns for the Stage/Freight into account in `startVerification`.

Reply:
I might be missing something here.

Here's how I thought this would work:

1. If the annotation is present and matched current verification info then blank that info out and put the phase back to Verifying. Early return.

2. Clear the annotations only if we successfully updated State status. Meaning:

    1. On success, next reconcile will lack the annotations and also have blank verification info and phase Verifying -- it will look like we're starting fresh.

    2. On failure, next reconcile will still have the annotations and will try again to blank out the verification info and set status to Verifying.

Is the issue right now mainly that we're not clearing the annotations _conditionally_ on status update success?

Reply:
I think this is the only thing that may be blocking. Everything else lgtm.

Reply:
The problem is that if we blank out the verification information, and then return to have the annotation removed. `startVerification` is designed to look for existing AnalysisRuns _before_ creating a new one: https://github.com/akuity/kargo/blob/f6fd5d15daa6e1688c282bb88a91dba331e16480/internal/controller/stages/verification.go#L35-L71

With an early return, we do not have anything at hand which allows us to differentiate as the annotation has been removed. Given this, I opted for not returning early. Which allows us to conditionally skip looking for existing AnalysisRuns.

Reply:
> and then return to have the annotation removed. startVerification is designed to look for existing AnalysisRuns before creating a new one

I'd forgotten about this. That was to account for the scenario that we successfully launched an AnalysisRun and _then_ encountered an error updating status...

That would have safeguarded against duplicate runs being created.

Problem understood, but I guess I am still a little confused by how this _does_ work now. 👀 some more. Standby.

Reply:
Ok... I get it now.

We don't need a second reconciliation attempt in order for the new verification attempt to proceed. After you've blanked out the old into, the _current_ reconcilation attempt proceeds with starting verification anew.

There might be an edge case here... I am still 👀

Reply:
Continuing this convo in a more [relevant spot](https://github.com/akuity/kargo/pull/1611/files#r1526446585)